webpack/generateDetails.js

```diff
@@ -0,0 +1,168 @@
+/* eslint-env node */
+/* eslint-disable no-useless-escape */
+// NOTE: double escapes are needed for file path delimiters in webpack regexes for cross-platform support
+
+const fs = require('fs');
+const path = require('path');
+const appPaths = require('./appPaths');
+
+const scanTemplates = function (dir, entryRegex, allAppPaths, details, isProdMode) {
+    /**
+     * This recursively scans all template files within a given directory, looking
+     * for a metch to the `entryRegex` specified above. The first group this entry regex
+     * (`match[1]`) should return the webpack entry name, eg `hqwebapp/js/some_page`.
+     *
+     * As they are discovered, entries are appended to the `details.entries` dictionary.
+     * Additionally, `details.aliases` is updated when an app alias is first found,
+     * eg `hqwebapp/js` (the first two folders in the entry name).
+     * Lastly, the (unique) application name is added to `details.appsWithEntries`
+     *
+     * @param dir - path to directory to begin scanning
+     * @param entryRegex -  regex for identifying an entry in an .html page (see examples below)
+     * @param allAppPaths (dict) - keys are all the applications being scanned
+     *                             and values are paths to each application's folder
+     * @param details (dict) - the dictionary that will be modified by this function, formatted as follows:
+     *                          {
+     *                              entries: {},
+     *                              aliases: {},
+     *                              appsWithEntries: [],
+     *                          }
+     * @param isProdMode (boolean) - `true` if this should be run in production mode, the difference
+     *                          being that entry filenames end in `.[contenthash].js`, which is necessary for
+     *                          cache busting on production.
+     */
+    const files = fs.readdirSync(dir);
+
+    files.forEach((file) => {
+        let fullPath = path.join(dir, file);
+        let stats = fs.statSync(fullPath);
+
+        if (stats.isDirectory()) {
+            // Make sure we recursively scan subdirectories
+            scanTemplates(fullPath, entryRegex, allAppPaths, details, isProdMode);
+        } else if (stats.isFile() && fullPath.endsWith('.html')) {
+            let content = fs.readFileSync(fullPath, 'utf-8');
+            let match;
+
+            // Extract all matches of the {% webpack_main "path" %} tag
+            while ((match = entryRegex.exec(content)) !== null) {
+                let entryName = match[1];
+                let folders = entryName.split('/');
+                let appName = folders[0];
+                let fullEntryPath = path.join(allAppPaths[appName], 'static', `${entryName}.js`);
+
+                if (!fs.existsSync(fullEntryPath)) {
+                    console.warn(`JavaScript file not found: ${fullEntryPath}`);
```

Comment:
Should this be a hard error? What are the implications of hitting this condition?